### PR TITLE
Support Push/Delete to http server and file system

### PR DESCRIPTION
**Pull request by yugangw-msft**
1. The missing piece is to support V3 repository style when the target is a file system.
2. The code around "HttpHandlerResourceV2" and "STSAuthHelper.cs" are temporary and will be gone in near future.

EDIT: CI passed

### CODE REVIEW

**Comment by yishaigalatzer** [comment]
CC @emgarten @deepakaravindr @anurse @zhili1208

File: src/NuGet.Clients/NuGet.CommandLine/Common/ConsoleCredentialProvider.cs [hunk 1]
```diff
@@ -35,6 +35,11 @@ public ConsoleCredentialProvider(IConsole console)
             {
                 throw new ArgumentNullException(nameof(uri));
             }
+            if (nonInteractive)
```
Comment (emgarten) [inline src/NuGet.Clients/NuGet.CommandLine/Common/ConsoleCredentialProvider.cs:39]: There should be an empty line above this.

File: src/NuGet.Core/NuGet.Protocol.Core.Types/Push/PackageUpdater.cs [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+
+namespace NuGet.Protocol.Core.Types
+{
+    public class PackageUpdater
+    {
+        private const string ServiceEndpoint = "/api/v2/package";
+        private const string ApiKeyHeader = "X-NuGet-ApiKey";
+
+        //TODO remove these fields, when we have a consolidated common http handler resource to use.
+        private readonly Func<Task<HttpHandlerResource>> _messageHandlerFactory;
```
Comment (emgarten) [inline src/NuGet.Core/NuGet.Protocol.Core.Types/Push/PackageUpdater.cs]: It should be possible to remove these now

**Comment by emgarten** [comment]
NuGet.Protocol.Core.v2 should not be used, this project will be deleted soon. V2 here means Legacy NuGet.Core resources, not OData V2.

The resources can still be called V2 but they go into NuGet.Protocol.Core.v3. Take a look at the FindPackageById resources which have v2 implementations in the V3 project.

Everything needed for STS auth and the handler resource are already in NuGet.Protocol.Core.v3 so this code should not need to be copied.

There have been recent changes to clean up HttpSource. Instead of getting an httpmessagehandler resource you can now get the HttpSourceResource which will give you the single instance HttpSource for the endpoint.  HttpSource.Create(source) can be used also.

**Comment by emgarten** [comment]
:watch:

**Comment by yugangw-msft** [comment]
CC @emgarten please review.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs [hunk 3]
```diff
@@ -77,7 +77,8 @@ public void PushCommand_PushToFileSystemSourceUnixStyle()
         public void PushCommand_PushToFileSystemSourceUncStyle()
         {
             // UNC only works in Windows. So skip this test if we're running on Unix,
-            if (Path.DirectorySeparatorChar == '/')
+            if (Environment.OSVersion.Platform == PlatformID.MacOSX ||
+                  Environment.OSVersion.Platform == PlatformID.Unix)
```
Comment (alpaix) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs]: Going forward I'd suggest using a platform trait to indicate platform specific tests like this

``` C#
[Fact]
[Trait("Platform", "WindowsOnly")]
public void PushCommand_PushToFileSystemSourceUncStyle()
{
...
}
```

Having the trait one can invoke xunit runner with `-ExcludeTraits "Platform=WindowsOnly"`

Comment (yugangw-msft) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs]: yep, this option looks good me. CC @deepakaravindr

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs [hunk 4]
```diff
@@ -336,13 +338,6 @@ public void PushCommand_PushToServerBasicAuth()
                     using (var server = new MockServer())
                     {
                         server.Listener.AuthenticationSchemes = AuthenticationSchemes.Basic;
-                        server.Get.Add("/nuget", r =>
```
Comment (yugangw-msft) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:339]: FYI, like i commented out on the redirect tests, it was agreed we don't need to deal with redirect support. So this check becomes invalid, and hence I removed

File: src/NuGet.Core/NuGet.Protocol.Core.v3/PushCommandResource.cs [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+        /// Pushes a package to a FileSystem.
+        /// </summary>
+        /// <param name="pathToPackage">The path of package to be pushed.</param>
+        private void PushPackageToFileSystem(string pathToPackage)
+        {
+            string root = _source.LocalPath;
+            PackageArchiveReader reader = new PackageArchiveReader(pathToPackage);
+            PackageIdentity packageIdentity = reader.GetIdentity();
+
+            //TODD: support V3 repo style if detect it is
```
Comment (yugangw-msft) [inline src/NuGet.Core/NuGet.Protocol.Core.v3/PushCommandResource.cs:132]: Likely i will not have time to include V3 support in the next PR, if indeed, i will log a bug

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs (hunk 4 above)
Comment (yishaigalatzer) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:339]: Why is this invalid? the httpclient is following redirects by default as far as I know

Comment (yugangw-msft) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:339]: Like i mentioned, we do not send out the "get" like nuget2 does to resolve the redirect link which has been removed by nuget server, so we have no request to record

Comment (yugangw-msft) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:339]: And i can log a tracking bug. Based on our conversation last Friday, we do not need to worry about this for now.

Comment (yishaigalatzer) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:339]: I think it is fine, lets file a tracking bug and immediately close it as won't fix with justification (that this was a hack for old fwlink nuget.org used to do).

Comment (yugangw-msft) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs:339]: done. https://github.com/NuGet/Home/issues/2064

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs (hunk 3 above)
Comment (alpaix) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetPushCommandTest.cs]: Opened a follow up issue https://github.com/NuGet/Home/issues/2065

**Comment by emgarten** [comment]
:rocket:

**Comment by yugangw-msft** [comment]
I merged the same change through #262 for simpler commit history.

**Pull request closed without merging**